### PR TITLE
HelpRank/Feed

### PR DESCRIPTION
**Pull request by raymondng**
Simple integration for reco clicked event. 

Local dev mode recommendation unfortunately still broken. Unit tests work but will need to deploy to prod to monitor & validate.

The current approach re-uses the search attribution mechanism: hit cache, etc. Once the discoveries are recorded, rekeep machinery kicks in.

reco keep event handling is not part of this PR. As discussed that might be a weak signal ... TBD

@stkem @yingjieMiao

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/curator/RecommendationUserAction.scala [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+object RecommendationUserAction {
+  object Delivered extends RecommendationUserAction("delivered")
+  object Clicked extends RecommendationUserAction("clicked")
+  object Kept extends RecommendationUserAction("kept")
+  object MarkedGood extends RecommendationUserAction("marked_good")
+  object MarkedBad extends RecommendationUserAction("marked_bad")
+  object Trashed extends RecommendationUserAction("trashed")
+  object ImprovementSuggested extends RecommendationUserAction("improvement_suggested")
+}
```
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/curator/RecommendationUserAction.scala:13]: mind removing the same class in `CuratorAnalytics`?

Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/curator/RecommendationUserAction.scala:13]: Oh it didn't? I blame it on idea :) Will do.

File: kifi-backend/modules/heimdal/app/com/keepit/commander/EventTrackingCommander.scala [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        if (keeperIds.isEmpty) {
+          log.infoP("no keepers; skipped")
+          Future.successful[Unit]()
+        } else {
+          log.infoP(s"keepers=${keeperIds.mkString(",")}")
+          val hitUUID = ExternalId[ArticleSearchResult]()
+          val ts = userEvent.time
+          val numKeepers = keeperIds.size
+          implicit val dca = TransactionalCaching.Implicits.directCacheAccess
+          kifiHitCache.get(KifiHitKey(userId, uriId)) match {
```
Comment (yingjieMiao) [inline kifi-backend/modules/heimdal/app/com/keepit/commander/EventTrackingCommander.scala]: I assume this cache is driven by 'click events', not just search 'hit' events?

**Comment by yingjieMiao** [comment]
data flow lgtm. as discussed in person, this will record a recommendation keep event ONLY IF there is a click event shortly before (e.g. within 1 hour). (If there is no click event found in the cache, this recommendation keep is considered as a weak signal)

@stkem

File: kifi-backend/modules/heimdal/app/com/keepit/commander/EventTrackingCommander.scala [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+object EventTrackingCommander {
+  implicit class HeimdalContextWrapper(underlying: HeimdalContext) {
+    // helpers/workarounds -- until we have a stronger contract for events
+    def getIdOpt[T](key: String): Option[Id[T]] = underlying.get[Double](key) map { id => Id[T](id.toLong) }
+    def getIdsOpt[T](key: String): Option[Seq[Id[T]]] = underlying.getSeq[Double](key).map { ids => ids.map(id => Id[T](id.toLong)) }
+  }
+}
+
+class EventTrackingCommander @Inject() (
```
Comment (stkem) [inline kifi-backend/modules/heimdal/app/com/keepit/commander/EventTrackingCommander.scala]: It looks to me like this is help rank specific, which makes the name `EventTrackingCommander` a bit misleading. How about `HelpRankTrackingCommander`?

**Comment by stkem** [comment]
made a small comment on naming. lgtm otherwise.
(I think the one hour window things is perfectly fine)

File: kifi-backend/modules/heimdal/app/com/keepit/commander/EventTrackingCommander.scala (hunk 3 above)
Comment (raymondng) [inline kifi-backend/modules/heimdal/app/com/keepit/commander/EventTrackingCommander.scala]: Sure thing.

File: kifi-backend/modules/heimdal/app/com/keepit/commander/EventTrackingCommander.scala (hunk 2 above)
Comment (raymondng) [inline kifi-backend/modules/heimdal/app/com/keepit/commander/EventTrackingCommander.scala]: yes

**Comment by raymondng** [comment]
Thanks. Incorporated feedback.

**Pull request merged**